### PR TITLE
Make sure that all attention works the same

### PR DESCRIPTION
**Pull request by dirkgr**
Addresses https://github.com/allenai/allennlp/issues/5345.

* Added `ScaledDotProductMatrixAttention`, and converted the transformer toolkit to use it
* Added tests to ensure that all `Attention` and `MatrixAttention` implementations are interchangeable
* Fixed the signature of `ScaledDotProductAttention` to match the other `Attention` classes.

### CODE REVIEW

File: allennlp/modules/transformer/attention_module.py [hunk 1]
```diff
@@ -478,7 +473,7 @@ def __init__(
             attention_head_size=key_value_proj_dim,
             num_attention_heads=num_heads,
             output_linear=True,
-            scoring_func="scaled_dot_product",
+            scoring_func="dot_product",
```
Comment (AkshitaB) [inline allennlp/modules/transformer/attention_module.py:476]: Why do we change the default? The original transformer uses `scaled_dot_product`, right?

**Review by AkshitaB — APPROVED** [review]
LGTM. Just a question about the default scoring function.

File: allennlp/modules/transformer/attention_module.py (hunk 1 above)
Comment (dirkgr) [reply allennlp/modules/transformer/attention_module.py:476]: I asked you that on Slack. The original transformer uses `scaled_`, as per the paper, but in your implementation (which matches HF), the scaling factor is forced to 1, so it doesn't scale at all. I continue to be confused about this.

File: allennlp/modules/transformer/attention_module.py [hunk 2]
```diff
@@ -487,8 +482,6 @@ def __init__(
             relative_attention_num_buckets=relative_attention_num_buckets,
         )
 
-        self.attn = Attention.by_name(self.scoring_func)(scaling_factor=1, normalize=False)
```
Comment (dirkgr) [inline allennlp/modules/transformer/attention_module.py:490]: @AkshitaB, this is where the scaling factor is forced to `1`.

File: allennlp/modules/transformer/attention_module.py (hunk 1 above)
Comment (AkshitaB) [reply allennlp/modules/transformer/attention_module.py:476]: Sorry, I missed that. From what I can see, scaling factor is set to 1 for `T5Attention`, not for regular `SelfAttention`. I believe original T5 does the same. By default, we set a scaling factor for regular `SelfAttention`.

File: allennlp/modules/transformer/attention_module.py (hunk 2 above)
Comment (AkshitaB) [reply allennlp/modules/transformer/attention_module.py:490]: This is for `T5Attention`.

File: allennlp/modules/transformer/attention_module.py (hunk 1 above)
Comment (dirkgr) [reply allennlp/modules/transformer/attention_module.py:476]: Right. This is also only for `T5Attention`. I believe I left it the same by default.

Comment (dirkgr) [reply allennlp/modules/transformer/attention_module.py:476]: So I guess it's just T5 being extra?

Comment (dirkgr) [reply allennlp/modules/transformer/attention_module.py:476]: I re-read section 2.1 of the T5 paper, and it doesn't mention this at all 🤷🏼‍♂️.

Comment (AkshitaB) [reply allennlp/modules/transformer/attention_module.py:476]: Yes, a whole lot of finicky little training details aren't mentioned in the 60+ pages paper. I think we were following the HF implementation.

**Pull request merged**